Add Electron multi-arch/cross-compilation support

dennisameling opened the pull request:
Now that ringrtc has switched to `neon`'s N-API backend, we can cross-compile to different architectures for Electron 🎉  This PR adds support for building ia32 and arm64. Please note that this still only supports building from a x64 host.

This work is related to https://github.com/signalapp/Signal-Desktop/issues/3745 and https://github.com/signalapp/Signal-Desktop/issues/4461.

You can test this by running `make electron NODEJS_ARCH=arm64`. Supported archs are `x64`, `ia32`, `arm64` (in line with NodeJS' archs), default is x64.

**Tested on:**
- [x] darwin x64 (build + yarn test)
- [x] darwin arm64 (build)
- [x] windows x64 (build + yarn test)
- [x] windows arm64 (build + yarn test)
- [x] windows ia32 (build)
- [x] linux x64 (build + yarn test)
- [x] linux arm64 (build)
- [x] linux ia32 (build)

## MacOS (darwin) arm64 / Apple Silicon
~MacOS (darwin) arm64 doesn't seem to work yet~ **FIXED** ✔️: https://github.com/signalapp/ringrtc/pull/12#issuecomment-830689652

jim-signal commented:
Hi @dennisameling, thanks for working through this. Yes, as you know, we currently use neon for the FFI from Javascript to Rust. I think as they mention, there were some difficulties with 32-bit support. Also, unfortunately, we don't use their 'N-API', although it is on our roadmap. It seems like an update is coming soon from them though... We will keep monitoring for that.

dennisameling commented:
Neon 0.4.1 was released yesterday, which brings us a bit closer to cross-compilation support, but it still doesn't work with the NAN-based runtime unfortunately:

```
denni@DESKTOP-HIGHLVU MINGW64 ~/repos/ringrtc/src/rust (add-windows-multi-arch-support)
$  npm_config_arch=x86 npm_config_target_arch=x86 npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=8.2.5 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C link-arg=-s" cargo build --target i686-pc-windows-msvc --features electron --release
   Compiling neon-sys v0.4.1
   Compiling curve25519-dalek v2.1.0
   Compiling futures-executor v0.3.5
   Compiling futures v0.3.5
   Compiling x25519-dalek v0.6.0
error: failed to run custom build command for `neon-sys v0.4.1`

Caused by:
  process didn't exit successfully: `C:\Users\denni\repos\ringrtc\src\rust\target\release\build\neon-sys-52c875801be308ce\build-script-build` (exit code: 101)
--- stdout
'Skipping node-gyp installation as part of npm install.'
audited 97 packages in 1.561s
found 3 low severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
cargo:node_root_dir=C:\\Users\\denni\\.electron-gyp\\8.2.5
cargo:node_lib_file=C:\\\\Users\\\\denni\\\\.electron-gyp\\\\8.2.5\\\\<(target_arch)\\\\node.lib

--- stderr
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\denni\.cargo\registry\src\github.com-1ecc6299db9ec823\neon-sys-0.4.1\build.rs:98:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `neon-sys v0.4.1`

Caused by:
  process didn't exit successfully: `C:\Users\denni\repos\ringrtc\src\rust\target\release\build\neon-sys-52c875801be308ce\build-script-build` (exit code: 101)
--- stdout
'Skipping node-gyp installation as part of npm install.'
added 97 packages from 70 contributors and audited 97 packages in 4.663s
found 3 low severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
cargo:node_root_dir=C:\\Users\\denni\\.electron-gyp\\8.2.5
cargo:node_lib_file=C:\\\\Users\\\\denni\\\\.electron-gyp\\\\8.2.5\\\\<(target_arch)\\\\node.lib

--- stderr
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\denni\.cargo\registry\src\github.com-1ecc6299db9ec823\neon-sys-0.4.1\build.rs:98:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Let's just wait for the Neon time to release the N-API backend :) They are currently working on preparing a [transition plan](https://github.com/neon-bindings/neon/issues/596) for current NAN users 🚀

dennisameling commented on Makefile:
```diff
@@ -17,6 +17,10 @@ ANDROID_TARGETS := $(foreach t, $(BUILD_TYPES),     \
 
+# These can be overriden on the command line, e.g. "make electron TARGET_ARCH=x86 CARGO_ARCH=i686-pc-windows-msvc"
+TARGET_ARCH := x64
+CARGO_ARCH := x86_64-pc-windows-msvc
```
Note to self: replace this with the cargo arch of the host (Windows/MacOS/etc.)

jim-signal commented:
RingRTC has been updated to use the n-api runtime for the neon bindings, so hopefully you will finally be unblocked.

dennisameling commented:
This is fantastic news, thanks for the update! Give me a few days to look into this 😊🚀

dennisameling commented:
First results looking good! Was able to generate an `arm64` .node file alongside the `x64` one. `x86` led to linker errors though, will look into that later this week. Great start though!! To be continued

![image](https://user-images.githubusercontent.com/17739158/107571983-efd2fa80-6beb-11eb-8a8b-bcaceb486ddf.png)

dennisameling commented on bin/build-electron:
```diff
@@ -104,22 +122,22 @@ export MACOSX_DEPLOYMENT_TARGET="10.10"
+        npm_config_arch=${TARGET_ARCH} npm_config_target_arch=${TARGET_ARCH} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=11.2.0 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C target-feature=+crt-static" cargo build --target ${CARGO_TARGET} --features electron
     else
-        npm_config_arch=x64 npm_config_target_arch=x64 npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=11.2.0 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C link-arg=-s" cargo build --features electron --release
+        npm_config_arch=${TARGET_ARCH} npm_config_target_arch=${TARGET_ARCH} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=11.2.0 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C link-arg=-s -C target-feature=+crt-static" cargo build --target ${CARGO_TARGET} --features electron --release
```
Had to add `target-feature=+crt-static`, because the `ia32` build was throwing linker errors for `msvcrt`. According to a Rust RFC it is supported to add `target-feature=+crt-static`: https://github.com/rust-lang/rfcs/blob/master/text/1721-crt-static.md#customizing-linkage-to-the-c-runtime

dennisameling commented:
@jim-signal it's working! Have all three archs for Windows working now:

![image](https://user-images.githubusercontent.com/17739158/107663446-adf99100-6c8b-11eb-9105-f06866f61858.png)

It can be built using `make electron TARGET_ARCH=arm64` (x64 is the default)

Have created an example branch for testing here: https://github.com/dennisameling/signal-ringrtc-node/tree/cross-compilation-support. Tested on x64 and arm64, in both cases `npm test` is passing 🚀 

It looks like CI / Lints is failing because of the following error, do you have any tips?

> Error: Unable to create clippy annotations! Reason: HttpError: Resource not accessible by integration
Warning: It seems that this Action is executed from the forked repository.
Warning: GitHub Actions are not allowed to create Check annotations, when executed for a forked repos. See https://github.com/actions-rs/clippy-check/issues/2 for details.

Planning to test cross-compilation on Linux and MacOS later today.

dennisameling commented:
Linux + MacOS x64 also working after this change 🚀  Reported my findings in https://github.com/signalapp/ringrtc/pull/12#issue-486201934

jim-signal commented on bin/build-electron:
```diff
@@ -56,31 +59,50 @@ while [ "$1" != "" ]; do
+        DEFAULT_PLATFORM="win32"
+        CARGO_TARGET="${CARGO_ARCH}-pc-windows-msvc"
+        # Static linking to prevent build errors on Windows ia32
+        RUSTFLAGS_WIN="-C target-feature=+crt-static"
```
Could/should this be done in the src/rust/build.rs file? There is a section in there where we force it to link to the static libc.

dennisameling replied:
@jim-signal I saw that there's already something for that in `build.rs`:

https://github.com/signalapp/ringrtc/blob/8c9259ff7d7cae08a1c1714167e7c9d5e085f3be/src/rust/build.rs#L45-L50

When I don't add `-C target-feature=+crt-static` to the build process, I get linker errors:

```
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x86\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LARGEADDRESSAWARE" "/SAFESEH" "/LIBPATH:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.0.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.1.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.10.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.11.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.12.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.13.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.14.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.15.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.2.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.3.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.4.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.5.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.6.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.7.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.8.rcgu.o" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.ringrtc.cvmzzsmy-cgu.9.rcgu.o" "/OUT:C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.dll" "/DEF:C:\\Users\\Dennis\\AppData\\Local\\Temp\\rustcvmeYrS\\lib.def" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.2b2wqrt5ttopb0e9.rcgu.o" "/OPT:REF,ICF" "/DLL" "/IMPLIB:C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\ringrtc.dll.lib" "/DEBUG" "/NATVIS:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps" "/LIBPATH:C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\release\\deps" "/LIBPATH:../../src/webrtc/src/out/Release/obj/" "/LIBPATH:C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "webrtc.lib" "winmm.lib" "dmoguids.lib" "msdmo.lib" "wmcodecdspuuid.lib" "secur32.lib" "iphlpapi.lib" "libcmt.lib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libbase64-dec81ab446f62e09.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\liblibc-1c1e89aadc6b9fb0.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libneon-0686d6bc0dd0d6ba.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libsemver-b5ae68705386b0f0.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libsemver_parser-7c05dc7df11561ec.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcslice-ecdaae8bdff1a339.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libneon_runtime-f5a25469333a2517.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\liblibloading-d1ba61772e48b9f6.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libwinapi-8e9f49b6bd285354.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libsmallvec-fd14e9653beac749.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libtokio-29fa05a912bf2d41.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libnum_cpus-5ada646cd69bd0a3.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libpin_project_lite-4ad6b54f6b3c6b34.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libregex-7a86961ed5bfb7ef.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libthread_local-7e3b5a81c9e88937.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libregex_syntax-a7f96430cd6b6ad4.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libaho_corasick-13ce9f99e015b04f.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\liblazy_static-8adbd3d5ac795c6f.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libserde_json-fa0f04708dc4d332.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libryu-54a417814c329448.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libitoa-306b9ae59e583c94.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libserde-3e6d1279027d62cc.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libthiserror-cebe39f8c3e54657.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libhmac-519c517e9b74e166.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcrypto_mac-22c60dba3b034482.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libaes_ctr-d5a5a550793b9405.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libctr-18dd290096b381e7.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libaes_soft-6018baf98414cdfb.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcipher-ec40be554c9dcc71.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libsha2-75ce678b6df9a846.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcpuid_bool-ed065f766669b9b3.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libopaque_debug-b81c03213ec16149.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libblock_buffer-abe4ac60febb6928.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\librand-2099ab509b662c09.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\librand_chacha-fcbf1f0bdda88daa.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libppv_lite86-617c2dbdcaa3b30d.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libhkdf-5e22431a56356815.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libhmac-7eacb58a2d12a71b.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcrypto_mac-a4c89d03545e5fbb.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libdigest-ef89ad212bb6e9d8.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libgeneric_array-ba9d0da38ac4b07d.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libprost-902e37477f2cbffa.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libbytes-0c5b701a5b048f9a.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libx25519_dalek-2c45ffce1483d0c8.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcurve25519_dalek-6c765d3bf5d9acff.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libsubtle-4c476c8ec3b8559f.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libzeroize-92b5a284b0d5c763.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\librand_core-6abb38f769da791e.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libgetrandom-ff0dd3aa737217c3.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libdigest-de83356e5cae0a2e.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libgeneric_array-1d29f4de1cb1a3a2.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libtypenum-3695c9289495ef66.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libbyteorder-29142b0ab8bd55a4.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libstatic_assertions-0abfed3606d79f43.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\liblog-cdd0cfede6239e7b.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcfg_if-ad7ef0d6cbae1fed.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures-104ae9cbd5723bde.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_executor-7d3510d420977ff8.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_util-b9e8ab39d0483d85.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libmemchr-0e660e3abdafe1d1.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libproc_macro_nested-5f39e20fbfe09011.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_io-90ce7ac3aa792b1b.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libslab-7bc94dc3b903c652.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_channel-832d675866ac38ca.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libpin_project-7c93e5e3c7bde49f.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_sink-cf36a805deb031be.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_task-f777fe1f46435374.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libonce_cell-19ed83ac2e2d0b8d.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libpin_utils-82cde187427fa1a3.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfutures_core-c552bdeea4b6dc72.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libfailure-04fd84390c860ddd.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libbacktrace-b6af707aa65a01c3.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\libcfg_if-709f3978065f6469.rlib" "C:\\Users\\Dennis\\repos-c\\ringrtc\\src\\rust\\target\\i686-pc-windows-msvc\\release\\deps\\librustc_demangle-69b9d79268602572.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libstd-f4c158dab33a3977.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libpanic_unwind-a37722c46f2a18b4.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_demangle-1c6d276b48ed2ce8.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libhashbrown-9a87a105949005d7.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-78a58b3ad1e3d735.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libunwind-02743839f37c6208.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcfg_if-b228f18e39d8285c.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\liblibc-1b403106181c3862.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\liballoc-8342f74face2334c.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_std_workspace_core-d12772f22e8379ca.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcore-81bf76d233082673.rlib" "C:\\Users\\Dennis\\.rustup\\toolchains\\1.47.0-x86_64-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcompiler_builtins-90bdc0d502053891.rlib" "advapi32.lib" "cfgmgr32.lib" "gdi32.lib" "kernel32.lib" "msimg32.lib" "opengl32.lib" "user32.lib" "winspool.lib" "advapi32.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "msvcrt.lib" "-s"
  = note: LINK : warning LNK4044: unrecognized option '/s'; ignored
             Creating library C:\Users\Dennis\repos-c\ringrtc\src\rust\target\i686-pc-windows-msvc\release\deps\ringrtc.dll.lib and object C:\Users\Dennis\repos-c\ringrtc\src\rust\target\i686-pc-windows-msvc\release\deps\ringrtc.dll.exp
          msvcrt.lib(chandler4gs.obj) : error LNK2019: unresolved external symbol __except_handler4_common referenced in function __except_handler4
          C:\Users\Dennis\repos-c\ringrtc\src\rust\target\i686-pc-windows-msvc\release\deps\ringrtc.dll : fatal error LNK1120: 1 unresolved externals


error: aborting due to previous error

error: could not compile `ringrtc`.

To learn more, run the command again with --verbose.
make: *** [Makefile:74: electron] Error 101
```

However, when I remove the static linking logic from `build.rs`, it works as expected.

My knowledge of static/dynamic linking is very limited, but could it be that an "all or nothing" approach is needed? So either compile everything statically through `-C target-feature=+crt-static`, or dynamically by removing `cargo:rustc-link-lib=libcmt`? What would you advise?

jim-signal replied:
For me both of those settings are needed, and there is no way to set these flags in build.rs, so it seems okay.

dennisameling commented:
Hi @jim-signal, is there anything I can help with to push this over the finish line? This is one of the final bits I need to get a working build on Apple Silicon, and support for that platform is in growing demand: https://github.com/signalapp/Signal-Desktop/issues/4461#issuecomment-782869558

Thanks in advance!

ashwaniii commented:
Probably this is the bottleneck in Apple Silicon support of Signal.
Do we have any updates?

Be-ing commented:
@dennisameling there is now a conflict in bin/build-electron

jim-signal commented:
Hi @dennisameling If you can rebase on the latest, and squash your changes to one commit, we can cherry-pick it for a release. Sorry for the current delays.

dennisameling commented:
@jim-signal rebased 👍🏼 also updated the docs at `BUILDING.md` for the new cross-compilation support.

And some very good news: I was also able to build for Apple Silicon now, presumably thanks to the [recent updates](https://github.com/signalapp/webrtc/pull/48) in Signal's WebRTC fork 🎉

dennisameling commented:
Just found out it's also possible to cross-compile for Linux arm64 and ia32. Needed to do the following for that, similar to https://github.com/signalapp/zkgroup/pull/14:

- Instruct cargo to use the correct linkers (see `.cargo/config.toml`)
- Update neon from 0.7.0 to 0.7.1 since it [includes a fix for Linux arm64](https://github.com/neon-bindings/neon/pull/668) 

Then you can run the following to cross-compile for Linux arm64 and ia32 (tested on Ubuntu 20.04 x64):

```
sudo apt-get update && sudo apt-get install -y crossbuild-essential-arm64 crossbuild-essential-i386
make electron PLATFORM=unix NODEJS_ARCH=arm64
make electron PLATFORM=unix NODEJS_ARCH=ia32
```

Here's the native binaries I generated through this: https://github.com/dennisameling/signal-ringrtc-node/tree/cross-compilation-support-new/build/linux

jrose-signal commented on .cargo/config.toml:
```diff
@@ -0,0 +1,4 @@
+[target.aarch64-unknown-linux-gnu]
+linker = "aarch64-linux-gnu-gcc"
+[target.i686-unknown-linux-gnu]
+linker = "i686-linux-gnu-gcc"
```
Sorry, can I ask why these are necessary? Shouldn't cross-compiling always find an appropriate cross-linker? (And if you're *not* cross-compiling, shouldn't you use the default linker?)

dennisameling replied:
Without this, the build breaks because it will use the x64 linker. Also see https://wiki.pine64.org/index.php?title=Cross-compiling&mobileaction=toggle_view_desktop#cargo. This only applies to Linux, on Windows and macOS this is not needed.

Your comment about not cross-compiling makes sense, so what I could do is add some logic to `bin/build-electron` that if the toolchain is `x86_64-unknown-linux` (the logic is already there), and the target is `aarch64` or `i686`, it uses said linkers. What do you think about that?

jrose-signal replied:
I guess my hesitancy comes from writing down specific linkers *anywhere* in a user-invoked build process. If I use gold as my cross-linker rather than gcc, the build's going to fail here. What do you think about moving the setting to an [environment variable](https://doc.rust-lang.org/cargo/reference/config.html#targettriplelinker) in the GitHub workflow, rather than having it be a default in the build scripts?

jrose-signal commented on src/rust/Cargo.toml:
```diff
@@ -49,7 +49,7 @@ simplelog   = { version = "0.7", optional = true, default-features = false }
 
 # Optional, needed by the "electron" feature
-neon = { version = "0.7.0", optional = true, default-features = false, features = ["napi-1"] }
+neon = { version = "0.7.1", optional = true, default-features = false, features = ["napi-1"] }
```
Why the bump here?

dennisameling commented:
@jrose-signal Alright, as discussed in https://github.com/signalapp/zkgroup/pull/14#issuecomment-832348867 we will just do the absolute minimum work now for cross-compilation scenarios, so that folks at least can cross-compile themselves until the Signal team is ready for it. I just updated the PR accordingly and rebased. Hope the tiny `neon` version bump from 0.7.0 to 0.7.1 is OK as it's needed for folks who want to cross-compile on Linux 😊 thanks in advance!

jim-signal reviewed (approved):
@dennisameling This PR looks good. We are trying to incorporate it in an upcoming 2.10.0 release. Will let you know when it is taken and close it here. Thanks!

jim-signal commented:
This PR has been merged to this commit: 142cbdc43df92afd647bd436ba2a861340e8e9a0

Thank you!
